### ISSUE TITLE
Mapping a dataset squashes class_names from image_dataset_from_directory

### ISSUE THREAD

**Issue by jamesbraza** [tensorflow/tensorflow#56233]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.4.4

### Custom Code

Yes

### OS Platform and Distribution

n/a

### Mobile device

n/a

### Python version

3.8

### Bazel version

n/a

### GCC/Compiler version

n/a

### CUDA/cuDNN version

n/a

### GPU model and memory

n/a

### Current Behaviour?

Using [`image_dataset_from_directory`](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory) will correctly create a `Dataset` that contains the labels as `class_names`.

However, when you transform that dataset (e.g. via `.map`) the `class_names` get clobbered.

Ideally, `.map`'ing or `.batch`'ing a `Dataset` won't clobber `class_names`.

### Standalone code to reproduce the issue

Here's how to get the dataset (from [Image classification from scratch](https://keras.io/examples/vision/image_classification_from_scratch/)).

```shell
curl -O https://download.microsoft.com/download/3/E/1/3E1C3F21-ECDB-4869-8368-6DEBA77B919F/kagglecatsanddogs_5340.zip
unzip kagglecatsanddogs_5340.zip
```

Then running this:

```python
import tensorflow as tf

train_ds = tf.keras.preprocessing.image_dataset_from_directory(
    "PetImages",
    batch_size=32,
)
assert hasattr(train_ds, "class_names")  # Passes
mapped_train_ds = train_ds.map(lambda x, y: (x, y))
assert hasattr(mapped_train_ds, "class_names")  # Doesn't pass
```

### Relevant log output

_No response_</details>
